contracts/test/KyberElasticSampler.t.sol

```diff
@@ -0,0 +1,223 @@
+        poolPath[0] = ETH_KNC_POOL_100_BIP;
+        poolPath[1] = LINK_ETH_POOL_30_BIP;
+
+        testAllAmountsAndPathsForSells(tokenPath, poolPath);
+    }
+
+    function testAllAmountsAndPathsForSells(address[] memory tokenPath, IPool[] memory poolPath) private {
+        uint256 kyberQuoterGasUsage;
+        uint256 multiQuoterGasUsage;
+
+        bytes memory path = toPath(tokenPath, poolPath);
+        bytes memory reversePath = toPath(reverseTokenPath(tokenPath), reversePoolPath(poolPath));
+
+        console.log("Quoter Gas Comparison");
+        console.log("Token Path:");
+        for (uint256 i = 0; i < tokenPath.length; ++i) {
+            console.logAddress(address(tokenPath[i]));
+        }
+
+        for (uint256 i = 0; i < testAmounts.length; ++i) {
+            (kyberQuoterGasUsage, multiQuoterGasUsage) = compareQuoterSells(path, testAmounts[i]);
+            console.log(
+                "Normal Path Sell: test=%d, kyberQuoterGasUsage=%d, multiQuoterGasUsage=%d",
+                i + 1,
+                kyberQuoterGasUsage,
+                multiQuoterGasUsage
+            );
+            (kyberQuoterGasUsage, multiQuoterGasUsage) = compareQuoterSells(reversePath, testAmounts[i]);
+            console.log(
+                "Reverse Path Sell: test=%d, kyberQuoterGasUsage=%d, multiQuoterGasUsage=%d",
+                i + 1,
+                kyberQuoterGasUsage,
+                multiQuoterGasUsage
+            );
+            (kyberQuoterGasUsage, multiQuoterGasUsage) = compareQuoterBuys(path, testAmounts[i]);
+            console.log(
+                "Normal Path Buy: test=%d, kyberQuoterGasUsage=%d, multiQuoterGasUsage=%d",
+                i + 1,
+                kyberQuoterGasUsage,
+                multiQuoterGasUsage
+            );
+            (kyberQuoterGasUsage, multiQuoterGasUsage) = compareQuoterBuys(reversePath, testAmounts[i]);
+            console.log(
+                "Reverse Path Buy: test=%d, kyberQuoterGasUsage=%d, multiQuoterGasUsage=%d",
+                i + 1,
+                kyberQuoterGasUsage,
+                multiQuoterGasUsage
+            );
+        }
+    }
+
+    function compareQuoterSells(
+        bytes memory path,
+        uint256[] memory amountsIn
+    ) private returns (uint256 kyberQuoterGasUsage, uint256 multiQuoterGasUsage) {
+        uint256 gas0 = gasleft();
+        (uint256[] memory multiQuoterAmountsOut, ) = multiQuoter.quoteExactMultiInput(factory, path, amountsIn);
+        uint256 gas1 = gasleft();
+
+        for (uint256 i = 0; i < amountsIn.length; ++i) {
+            try kyberQuoter.quoteExactInput(path, amountsIn[i]) returns (
+                uint256 kyberQuoterAmountOut,
+                uint160[] memory /* sqrtPriceX96AfterList */,
+                uint32[] memory /* initializedTicksCrossedList */,
+                uint256 /* gasEstimate */
+            ) {
+                assertApproxEqRel(multiQuoterAmountsOut[i], kyberQuoterAmountOut, ERROR_THRESHOLD_10_BPS);
```

Comment:
Just curious, how low can you drop the error threshold before tests start to fail? Is the same rounding error for UniswapV3 present for Kyber?

Reply:
It's definitely higher than the errors in UniswapV3. I was seeing errors of < 1 bp for bigger trades ($100k with ~$.30 error). Errors for small trades ($2, $.01 error, so 5 bps) made it hard to get a tighter bound. Wanted to timebox the error investigation and possibly get a full simbot run before going back to it.